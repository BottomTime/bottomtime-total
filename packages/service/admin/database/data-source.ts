import path from 'path';
import { DataSource } from 'typeorm';

export async function getDataSource(
  url: string,
  requireSsl: boolean,
): Promise<DataSource> {
  const client = new DataSource({
    type: 'postgres',
    url,
    entities: [path.resolve(__dirname, '../../src/data/**/*.entity.ts')],
    ssl: requireSsl,
  });
  return await client.initialize();
}
