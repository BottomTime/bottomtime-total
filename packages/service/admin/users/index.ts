import { CommandModule } from 'yargs';

import { createAdmin } from './create-admin';
import { getUserToken } from './get-user-token';

export const userModule: CommandModule<{
  'postgres-uri': string;
  'require-ssl': boolean;
}> = {
  command: 'user',

  describe: 'Commands for managing a user account',

  builder(yargs) {
    return yargs
      .command(
        'new-admin',
        'Creates a new admin user.',
        (yargs) => {
          return yargs
            .positional('username', {
              alias: 'u',
              description: "The user's username.",
              type: 'string',
            })
            .positional('password', {
              alias: 'p',
              description: 'The password to set on the new admin account',
              type: 'string',
            })
            .help();
        },
        async (yargs) => {
          await createAdmin({
            postgresRequireSsl: yargs.requireSsl,
            potgresUri: yargs.postgresUri,
            username: yargs.username,
            password: yargs.password,
          });
        },
      )
      .command(
        'token <user>',
        'Gets a token for a user',
        (yargs) => {
          return yargs
            .positional('user', {
              demandOption: true,
              description: "The user's username or email address.",
              type: 'string',
            })
            .help();
        },
        async (yargs) => {
          await getUserToken(yargs.postgresUri, yargs.requireSsl, yargs.user);
        },
      )
      .help();
  },

  async handler() {
    /* TODO */
  },
};
