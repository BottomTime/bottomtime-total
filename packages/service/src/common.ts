import { z } from 'zod';
import { DepthUnit } from './constants';

export const BooleanString = z
  .enum(['true', 'false'])
  .transform((value) => value === 'true');

export const DepthSchema = z.object({
  depth: z.number().min(0),
  unit: z.nativeEnum(DepthUnit),
});
export type Depth = z.infer<typeof DepthSchema>;

export const GpsCoordinatesSchema = z.object({
  lat: z.number().gte(-90).lte(90),
  lon: z.number().gte(-180).lte(180),
});
export type GpsCoordinates = z.infer<typeof GpsCoordinatesSchema>;

export interface Range {
  min: number;
  max: number;
}
