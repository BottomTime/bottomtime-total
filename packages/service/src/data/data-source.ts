import path from 'path';
import { DataSource, DataSourceOptions } from 'typeorm';

import { Config } from '../config';

export const PostgresDataSourceOptions: DataSourceOptions = {
  type: 'postgres',
  url: Config.postgresUri,
  entities: [path.join(__dirname, './**/*.entity.ts')],
  synchronize: false,
};

export async function initDataSource(): Promise<DataSource> {
  const AppDataSource = new DataSource(PostgresDataSourceOptions);
  return await AppDataSource.initialize();
}
