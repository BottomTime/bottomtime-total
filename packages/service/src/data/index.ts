export * from './data-source';
export * from './get-query-vector';

export * from './certification.entity';
export * from './dive-site-review.entity';
export * from './dive-site.entity';
export * from './friend-request.entity';
export * from './friendship.entity';
export * from './tank.entity';
export * from './user-certification.entity';
export * from './user-oauth.entity';
export * from './user.entity';
