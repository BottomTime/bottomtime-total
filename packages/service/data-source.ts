import path from 'path';
import { DataSource } from 'typeorm';

import { PostgresDataSourceOptions } from './src/data/data-source';

export default new DataSource({
  ...PostgresDataSourceOptions,
  migrations: [path.join(__dirname, './migrations/*.ts')],
});
