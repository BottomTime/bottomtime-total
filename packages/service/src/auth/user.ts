import { PasswordResetTokenStatus, UserDTO, UserRole } from '@bottomtime/api';

import { ConflictException, Logger } from '@nestjs/common';

import { compare, hash } from 'bcrypt';
import { randomBytes } from 'crypto';
import dayjs from 'dayjs';
import { Repository } from 'typeorm';

import { DefaultUserSettings } from '../common';
import { Config } from '../config';
import { UserEntity } from '../data';
import { Profile } from './profile';

type UserDataSettings = NonNullable<
  Pick<
    UserEntity,
    'depthUnit' | 'pressureUnit' | 'temperatureUnit' | 'weightUnit'
  >
>;
export type UserSettings = Required<{
  [K in keyof UserDataSettings]: Exclude<UserDataSettings[K], null>;
}>;

export class User implements Express.User {
  private readonly log = new Logger(User.name);

  private _profile: Profile | undefined;

  constructor(
    private readonly Users: Repository<UserEntity>,
    private readonly data: UserEntity,
  ) {}

  get username(): string {
    return this.data.username;
  }

  get email(): string | null {
    return this.data.email;
  }

  get emailVerified(): boolean {
    return this.data.emailVerified;
  }

  get id(): string {
    return this.data.id;
  }

  get hasPassword(): boolean {
    return !!this.data.passwordHash;
  }

  get lastLogin(): Date | null {
    return this.data.lastLogin;
  }

  get lastPasswordChange(): Date | null {
    return this.data.lastPasswordChange;
  }

  get isLockedOut(): boolean {
    return this.data.isLockedOut;
  }

  get memberSince(): Date {
    return this.data.memberSince;
  }

  get profile(): Profile {
    if (this._profile) return this._profile;

    this._profile = new Profile(this.Users, this.data);
    return this._profile;
  }

  get role(): UserRole {
    return this.data.role as UserRole;
  }

  get settings(): UserSettings {
    return {
      depthUnit: this.data.depthUnit ?? DefaultUserSettings.depthUnit,
      pressureUnit: this.data.pressureUnit ?? DefaultUserSettings.pressureUnit,
      temperatureUnit:
        this.data.temperatureUnit ?? DefaultUserSettings.temperatureUnit,
      weightUnit: this.data.weightUnit ?? DefaultUserSettings.weightUnit,
    };
  }

  async changeSettings(settings: Partial<UserSettings>): Promise<void> {
    if (settings.depthUnit) this.data.depthUnit = settings.depthUnit;

    if (settings.pressureUnit) this.data.pressureUnit = settings.pressureUnit;

    if (settings.temperatureUnit)
      this.data.temperatureUnit = settings.temperatureUnit;

    if (settings.weightUnit) this.data.weightUnit = settings.weightUnit;

    await this.Users.save(this.data);
  }

  async changeUsername(newUsername: string): Promise<void> {
    const lowered = newUsername.toLowerCase();

    this.log.debug(`Attempting to change username for user with ID ${this.id}`);
    const exists = await this.Users.findOne({
      where: { usernameLowered: lowered },
    });
    if (exists) {
      throw new ConflictException(
        `Username "${newUsername}" is already taken.`,
      );
    }

    this.data.username = newUsername;
    this.data.usernameLowered = lowered;

    await this.Users.save(this.data);
  }

  async changeEmail(newEmail: string): Promise<void> {
    const lowered = newEmail.toLowerCase();

    this.log.debug(`Attempting to change email for user with ID ${this.id}`);
    const exists = await this.Users.findOne({
      where: { emailLowered: lowered },
    });
    if (exists) {
      throw new ConflictException(`Email "${newEmail}" is already taken.`);
    }

    this.data.email = newEmail;
    this.data.emailLowered = lowered;

    // Invalidate any existing email verification token so that if there is one floating
    // around in the wild it cannot be re-played to verify a fake email address.
    this.data.emailVerificationToken = null;
    this.data.emailVerificationTokenExpiration = null;

    await this.Users.save(this.data);
  }

  async requestEmailVerificationToken(): Promise<string> {
    const token = randomBytes(32).toString('base64url');

    this.data.emailVerificationToken = token;
    this.data.emailVerificationTokenExpiration = dayjs().add(2, 'day').toDate();
    await this.Users.save(this.data);

    return token;
  }

  async verifyEmail(token: string): Promise<boolean> {
    if (
      !this.data.emailVerificationToken ||
      !this.data.emailVerificationTokenExpiration
    ) {
      this.log.debug(
        `Email verification failed because user ${this.username} does not have a verification token set.`,
      );
      return false;
    }

    if (this.data.emailVerificationToken !== token) {
      this.log.debug(
        `Email verification failed for user ${this.username} because the proided token did not match.`,
      );
      return false;
    }

    if (this.data.emailVerificationTokenExpiration < new Date()) {
      this.log.debug(
        `Email verification for user ${this.username} failed because the token has expired.`,
      );
      return false;
    }

    this.data.emailVerified = true;
    this.data.emailVerificationToken = null;
    this.data.emailVerificationTokenExpiration = null;

    this.log.log(
      `User ${this.username} has verified their email address: ${this.email}.`,
    );
    await this.Users.save(this.data);

    return true;
  }

  async changePassword(
    oldPassword: string,
    newPassword: string,
  ): Promise<boolean> {
    if (!this.data.passwordHash) {
      return false;
    }

    this.log.debug(`Attempting to change password for user ${this.username}`);
    const oldPassMatches = await compare(oldPassword, this.data.passwordHash);
    if (!oldPassMatches) {
      this.log.debug(
        `Password change for user ${this.username} failed because the old password was incorrect.`,
      );
      return false;
    }

    const passwordHash = await hash(newPassword, Config.passwordSaltRounds);

    this.data.passwordHash = passwordHash;
    this.data.lastPasswordChange = new Date();
    await this.Users.save(this.data);

    this.log.log(`User ${this.username} has changed their password.`);

    return true;
  }

  async requestPasswordResetToken(): Promise<string> {
    const token = randomBytes(32).toString('base64url');
    this.data.passwordResetToken = token;
    this.data.passwordResetTokenExpiration = dayjs().add(2, 'day').toDate();

    await this.Users.save(this.data);

    return token;
  }

  validatePasswordResetToken(token: string): PasswordResetTokenStatus {
    if (
      !this.data.passwordResetToken ||
      !this.data.passwordResetTokenExpiration ||
      this.data.passwordResetToken !== token
    ) {
      return PasswordResetTokenStatus.Invalid;
    }

    if (this.data.passwordResetTokenExpiration < new Date()) {
      return PasswordResetTokenStatus.Expired;
    }

    return PasswordResetTokenStatus.Valid;
  }

  async resetPassword(token: string, newPassword: string): Promise<boolean> {
    this.log.debug(`Attempting to reset password for user ${this.username}`);

    if (
      this.data.passwordResetTokenExpiration! < new Date() ||
      token !== this.data.passwordResetToken
    ) {
      this.log.debug(
        `Password reset failed for user ${this.username}. No reset token was set.`,
      );
      return false;
    }

    const passwordHash = await hash(newPassword, Config.passwordSaltRounds);

    this.data.passwordHash = passwordHash;
    this.data.lastPasswordChange = new Date();
    this.data.passwordResetToken = null;
    this.data.passwordResetTokenExpiration = null;
    await this.Users.save(this.data);

    this.log.log(
      `User ${this.username} has successfully reset their password.`,
    );

    return true;
  }

  async updateLastLogin(): Promise<void> {
    this.data.lastLogin = new Date();
    await this.Users.save(this.data);
  }

  toJSON(): UserDTO {
    return {
      emailVerified: this.emailVerified,
      hasPassword: this.hasPassword,
      id: this.id,
      isLockedOut: this.isLockedOut,
      memberSince: this.memberSince,
      profile: this.profile.toJSON(),
      role: this.role,
      username: this.username,
      email: this.email ?? undefined,
      lastLogin: this.lastLogin ?? undefined,
      lastPasswordChange: this.lastPasswordChange ?? undefined,
      settings: this.settings,
    };
  }

  toEntity(): UserEntity {
    return { ...this.data };
  }
}
