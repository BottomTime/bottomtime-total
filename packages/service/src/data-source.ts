import 'reflect-metadata';
import { DataSourceOptions } from 'typeorm';

import { Config } from './config';
import * as Entities from './data';

export const PostgresDataSourceOptions: DataSourceOptions = {
  type: 'postgres',
  url: Config.postgresUri,
  entities: Object.values(Entities),
  synchronize: false,
};
