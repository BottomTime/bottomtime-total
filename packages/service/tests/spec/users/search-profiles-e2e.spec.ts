import { ProfileDTO, SortOrder, UsersSortBy } from '@bottomtime/api';

import { HttpServer, INestApplication } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';

import request from 'supertest';
import { Repository } from 'typeorm';

import { Queues } from '../../../src/common';
import { FriendshipEntity, UserEntity } from '../../../src/data';
import { QueueModule } from '../../../src/queue';
import { UsersService } from '../../../src/users';
import { UserController } from '../../../src/users/user.controller';
import { UsersController } from '../../../src/users/users.controller';
import { dataSource } from '../../data-source';
import TestUserData from '../../fixtures/user-search-data.json';
import { createAuthHeader, createTestApp, parseUserJSON } from '../../utils';
import { createTestFriendship } from '../../utils/create-test-friendship';

const SearchUrl = '/api/users';

describe('Searching Profiles E2E Tests', () => {
  let app: INestApplication;
  let server: HttpServer;
  let authHeader: [string, string];

  let Users: Repository<UserEntity>;
  let Friends: Repository<FriendshipEntity>;
  let users: UserEntity[];

  beforeAll(async () => {
    users = TestUserData.map((data) => parseUserJSON(data));
    authHeader = await createAuthHeader(users[0].id);
    app = await createTestApp({
      imports: [
        TypeOrmModule.forFeature([UserEntity]),
        QueueModule.forFeature({
          key: Queues.email,
          queueUrl: 'http://localhost:4566/000000000000/email-queue',
        }),
      ],
      providers: [UsersService],
      controllers: [UsersController],
    });
    server = app.getHttpServer();

    Users = dataSource.getRepository(UserEntity);
    Friends = dataSource.getRepository(FriendshipEntity);
  });

  beforeEach(async () => {
    await Users.save(users);
  });

  afterAll(async () => {
    await app.close();
  });

  it('will list profiles when no query string parameters are provided', async () => {
    const { body } = await request(server)
      .get(SearchUrl)
      .set(...authHeader)
      .query({ limit: 10 })
      .expect(200);

    expect(body.totalCount).toBe(100);
    expect(body.users).toHaveLength(10);
    expect(body.users.map((u: ProfileDTO) => u.username)).toMatchSnapshot();
  });

  it('will perform a text search for user profiles', async () => {
    const options = {
      query: 'ipsam',
    };
    const { body: result } = await request(server)
      .get(SearchUrl)
      .set(...authHeader)
      .query(options)
      .expect(200);

    expect(result.totalCount).toBe(28);
    expect(result.users.map((u: ProfileDTO) => u.username)).toMatchSnapshot();
  });

  it('will allow filtering out profiles for the current user', async () => {
    const friendRelations: FriendshipEntity[] = [
      createTestFriendship(users[0].id, 'cbbdc55c-4f64-4fdf-ac63-5c42cfd80607'),
      createTestFriendship('cbbdc55c-4f64-4fdf-ac63-5c42cfd80607', users[0].id),

      createTestFriendship(users[0].id, 'a966e53e-0d66-4dff-8f7c-2c3658eea2e4'),
      createTestFriendship('a966e53e-0d66-4dff-8f7c-2c3658eea2e4', users[0].id),
    ];
    await Friends.save(friendRelations);

    const options = {
      filterFriends: true,
      limit: 15,
    };
    const { body: result } = await request(server)
      .get(SearchUrl)
      .set(...authHeader)
      .query(options)
      .expect(200);

    expect(result.totalCount).toBe(97);
    expect(
      result.users.map((u: ProfileDTO) => ({
        id: u.userId,
        username: u.username,
      })),
    ).toMatchSnapshot();
  });

  [
    {
      sortBy: UsersSortBy.MemberSince,
      sortOrder: SortOrder.Ascending,
    },
    {
      sortBy: UsersSortBy.MemberSince,
      sortOrder: SortOrder.Descending,
    },
    {
      sortBy: UsersSortBy.Username,
      sortOrder: SortOrder.Ascending,
    },
    {
      sortBy: UsersSortBy.Username,
      sortOrder: SortOrder.Descending,
    },
  ].forEach(({ sortBy, sortOrder }) => {
    it(`will sort results by ${sortBy} in ${sortOrder} order`, async () => {
      const { body } = await request(server)
        .get(SearchUrl)
        .set(...authHeader)
        .query({ sortBy, sortOrder })
        .expect(200);

      expect(body.totalCount).toBe(100);
      expect(
        body.users.map((u: ProfileDTO) => ({
          username: u.username,
          memberSince: u.memberSince,
        })),
      ).toMatchSnapshot();
    });
  });

  it('will return an empty result set if the query does not match any profiles', async () => {
    const options = {
      query: 'asdfasdfasdf',
    };
    const { body: result } = await request(server)
      .get(SearchUrl)
      .set(...authHeader)
      .query(options)
      .expect(200);

    expect(result.totalCount).toBe(0);
    expect(result.users).toHaveLength(0);
  });

  it('will return a 400 response if the query string is invalid', async () => {
    const { body: error } = await request(server)
      .get(SearchUrl)
      .set(...authHeader)
      .query({ sortBy: 'asdf', sortOrder: 17, query: true, limit: -1 })
      .expect(400);
    expect(error.details).toMatchSnapshot();
  });

  it('will return a 401 response if the user is not authenticated', async () => {
    await request(server).get(SearchUrl).expect(401);
  });
});
