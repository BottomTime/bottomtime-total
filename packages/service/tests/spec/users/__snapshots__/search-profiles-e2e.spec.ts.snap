// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Searching Profiles E2E Tests will allow admins to list all profiles, including private 1`] = `
[
  "Abigayle.Jaskolski61",
  "Adalberto.Rau18",
  "Adelbert93",
  "Admin",
  "Alba.Keeling",
  "Alessia.Dietrich4",
  "Alicia.Lebsack68",
  "Amara.Ernser14",
  "Amina2",
  "Anna99",
  "Antonette.Dickens0",
  "Assunta.Kautzer",
  "Assunta61",
  "Augustine.Bartoletti",
  "Augustus.Schneider32",
  "Aurelia.Walsh",
  "Aurore.Tillman",
  "Aurore68",
  "Bartholome.Beer37",
  "Benedict.Stoltenberg",
  "Bernardo16",
  "Bernie67",
  "Berniece.Bayer",
  "Bette.Kuhn84",
  "Bo.Douglas49",
  "Brittany3",
  "Caitlyn.Zieme",
  "Carissa.Dooley",
  "Carlee.Skiles",
  "Carmen.Von13",
]
`;

exports[`Searching Profiles E2E Tests will allow anonymous users to list public profiles 1`] = `
[
  "Abigayle.Jaskolski61",
  "Amina2",
  "Assunta.Kautzer",
  "Aurore.Tillman",
  "Aurore68",
  "Bartholome.Beer37",
  "Bernardo16",
  "Bernie67",
  "Brittany3",
  "Casper.Hagenes5",
  "Chad.Turcotte17",
  "Clifton.Powlowski",
  "Dejah.Bechtelar33",
  "Dorian60",
  "Dorthy34",
  "Dwight90",
  "Elian33",
  "Emerson.Nikolaus72",
  "Ernesto96",
  "Euna86",
  "Felipe35",
  "Georgianna9",
  "Golda.Stark",
  "Herminio44",
  "Houston.Mante",
]
`;

exports[`Searching Profiles E2E Tests will list profiles when no query string parameters are provided 1`] = `
[
  "Abigayle.Jaskolski61",
  "Alicia.Lebsack68",
  "Amara.Ernser14",
  "Amina2",
  "Anna99",
  "Antonette.Dickens0",
  "Assunta.Kautzer",
  "Assunta61",
  "Aurelia.Walsh",
  "Aurore.Tillman",
]
`;

exports[`Searching Profiles E2E Tests will perform a text search for user profiles 1`] = `
[
  "Kendra.Gusikowski39",
  "Shad.Daniel29",
  "Novella.Reichel",
  "Sallie.Schultz",
]
`;

exports[`Searching Profiles E2E Tests will return a 400 response if the query string is invalid 1`] = `
{
  "issues": [
    {
      "code": "invalid_enum_value",
      "message": "Invalid enum value. Expected 'username' | 'memberSince', received 'asdf'",
      "options": [
        "username",
        "memberSince",
      ],
      "path": [
        "sortBy",
      ],
      "received": "asdf",
    },
    {
      "code": "invalid_enum_value",
      "message": "Invalid enum value. Expected 'asc' | 'desc', received '17'",
      "options": [
        "asc",
        "desc",
      ],
      "path": [
        "sortOrder",
      ],
      "received": "17",
    },
    {
      "code": "too_small",
      "exact": false,
      "inclusive": false,
      "message": "Number must be greater than 0",
      "minimum": 0,
      "path": [
        "limit",
      ],
      "type": "number",
    },
  ],
}
`;

exports[`Searching Profiles E2E Tests will sort results by memberSince in asc order 1`] = `
[
  "Magali.Friesen",
  "Georgianna9",
  "Houston.Mante",
  "Hiram.Renner1",
  "Roxanne58",
  "Winfield.Ebert",
  "Reuben.Langosh45",
  "Emmet.Rogahn58",
  "Kaitlin64",
  "Erik.Adams93",
  "Nikki.Murphy99",
  "Carissa.Dooley",
  "Dorian60",
  "Ernesto96",
  "Aurore.Tillman",
  "Lolita56",
  "Rosamond56",
  "Kristy54",
  "Lina.Hayes98",
  "Brittany3",
  "Antonette.Dickens0",
  "Chad.Turcotte17",
  "Euna86",
  "Davion90",
  "Emmitt4",
  "Josie.Sawayn44",
  "Kendrick.Lakin",
  "Jerald.Klein37",
  "Nicolas.Keeling",
  "Simone.Howe10",
  "Bartholome.Beer37",
  "Casper.Hagenes5",
  "Clifton.Powlowski",
  "Jenifer.Mitchell92",
  "Glenda.Gleichner",
  "Winifred.Feil",
  "Kylie64",
  "Selmer68",
  "Josianne54",
  "Novella.Reichel",
  "Sheridan.Kulas13",
  "Felipe35",
  "Elian33",
  "Anna99",
  "Van89",
  "Kristoffer.Feest",
  "Katrine.Kiehn69",
  "Immanuel.Davis12",
  "Sven.Schulist13",
  "Kelvin5",
  "Shad.Daniel29",
  "Velda42",
  "Kenna.Dare",
  "Tyra.Rutherford",
  "Sallie.Schultz",
  "Stephen.Goldner60",
  "Assunta61",
  "Golda.Stark",
  "Mellie.Berge79",
  "Aurelia.Walsh",
  "Imelda1",
  "Jake.Hermann65",
  "Aurore68",
  "Henriette14",
  "Elroy67",
  "Keanu.Klocko",
  "Amina2",
  "Hilda.Hodkiewicz",
  "Bernie67",
  "Zora.Stroman31",
  "Herminio44",
  "Hortense20",
  "Carson.Auer",
  "Lexie76",
  "Kendra.Gusikowski39",
  "Ernestina0",
  "Assunta.Kautzer",
  "Jace88",
  "Samanta.Cummings",
  "Julia.Little89",
  "Skye.Walter59",
  "Emerson.Nikolaus72",
  "Abigayle.Jaskolski61",
  "Maiya37",
  "Judson.Halvorson41",
  "Alicia.Lebsack68",
  "Maverick.Bauch14",
  "Yesenia.Wisozk",
  "Amara.Ernser14",
  "Leopoldo39",
  "Dwight90",
  "Elisha.Erdman",
  "Cora91",
  "Dejah.Bechtelar33",
  "Rahul.Feil",
  "Talia.Parisian94",
  "Natalie.Corwin",
  "Jaunita71",
  "Dorthy34",
  "Bernardo16",
]
`;

exports[`Searching Profiles E2E Tests will sort results by memberSince in desc order 1`] = `
[
  "Leopold59",
  "Bernardo16",
  "Dorthy34",
  "Jaunita71",
  "Natalie.Corwin",
  "Talia.Parisian94",
  "Rahul.Feil",
  "Dejah.Bechtelar33",
  "Cora91",
  "Elisha.Erdman",
  "Dwight90",
  "Leopoldo39",
  "Amara.Ernser14",
  "Yesenia.Wisozk",
  "Maverick.Bauch14",
  "Alicia.Lebsack68",
  "Judson.Halvorson41",
  "Maiya37",
  "Abigayle.Jaskolski61",
  "Emerson.Nikolaus72",
  "Skye.Walter59",
  "Julia.Little89",
  "Samanta.Cummings",
  "Jace88",
  "Assunta.Kautzer",
  "Ernestina0",
  "Kendra.Gusikowski39",
  "Lexie76",
  "Carson.Auer",
  "Hortense20",
  "Herminio44",
  "Zora.Stroman31",
  "Bernie67",
  "Hilda.Hodkiewicz",
  "Amina2",
  "Keanu.Klocko",
  "Elroy67",
  "Henriette14",
  "Aurore68",
  "Jake.Hermann65",
  "Imelda1",
  "Aurelia.Walsh",
  "Mellie.Berge79",
  "Golda.Stark",
  "Assunta61",
  "Stephen.Goldner60",
  "Sallie.Schultz",
  "Tyra.Rutherford",
  "Kenna.Dare",
  "Velda42",
  "Shad.Daniel29",
  "Kelvin5",
  "Sven.Schulist13",
  "Immanuel.Davis12",
  "Katrine.Kiehn69",
  "Kristoffer.Feest",
  "Van89",
  "Anna99",
  "Elian33",
  "Felipe35",
  "Sheridan.Kulas13",
  "Novella.Reichel",
  "Josianne54",
  "Selmer68",
  "Kylie64",
  "Winifred.Feil",
  "Glenda.Gleichner",
  "Jenifer.Mitchell92",
  "Clifton.Powlowski",
  "Casper.Hagenes5",
  "Bartholome.Beer37",
  "Simone.Howe10",
  "Nicolas.Keeling",
  "Jerald.Klein37",
  "Kendrick.Lakin",
  "Josie.Sawayn44",
  "Emmitt4",
  "Davion90",
  "Euna86",
  "Chad.Turcotte17",
  "Antonette.Dickens0",
  "Brittany3",
  "Lina.Hayes98",
  "Kristy54",
  "Rosamond56",
  "Lolita56",
  "Aurore.Tillman",
  "Ernesto96",
  "Dorian60",
  "Carissa.Dooley",
  "Nikki.Murphy99",
  "Erik.Adams93",
  "Kaitlin64",
  "Emmet.Rogahn58",
  "Reuben.Langosh45",
  "Winfield.Ebert",
  "Roxanne58",
  "Hiram.Renner1",
  "Houston.Mante",
  "Georgianna9",
]
`;

exports[`Searching Profiles E2E Tests will sort results by username in desc order 1`] = `
[
  "Zora.Stroman31",
  "Yesenia.Wisozk",
  "Winifred.Feil",
  "Winfield.Ebert",
  "Velda42",
  "Van89",
  "Tyra.Rutherford",
  "Talia.Parisian94",
  "Sven.Schulist13",
  "Stephen.Goldner60",
  "Skye.Walter59",
  "Simone.Howe10",
  "Sheridan.Kulas13",
  "Shad.Daniel29",
  "Selmer68",
  "Samanta.Cummings",
  "Sallie.Schultz",
  "Roxanne58",
  "Rosamond56",
  "Reuben.Langosh45",
  "Rahul.Feil",
  "Novella.Reichel",
  "Nikki.Murphy99",
  "Nicolas.Keeling",
  "Natalie.Corwin",
  "Mellie.Berge79",
  "Maverick.Bauch14",
  "Maiya37",
  "Magali.Friesen",
  "Lolita56",
  "Lina.Hayes98",
  "Lexie76",
  "Leopoldo39",
  "Leopold59",
  "Kylie64",
  "Kristy54",
  "Kristoffer.Feest",
  "Kenna.Dare",
  "Kendrick.Lakin",
  "Kendra.Gusikowski39",
  "Kelvin5",
  "Keanu.Klocko",
  "Katrine.Kiehn69",
  "Kaitlin64",
  "Julia.Little89",
  "Judson.Halvorson41",
  "Josie.Sawayn44",
  "Josianne54",
  "Jerald.Klein37",
  "Jenifer.Mitchell92",
  "Jaunita71",
  "Jake.Hermann65",
  "Jace88",
  "Immanuel.Davis12",
  "Imelda1",
  "Houston.Mante",
  "Hortense20",
  "Hiram.Renner1",
  "Hilda.Hodkiewicz",
  "Herminio44",
  "Henriette14",
  "Golda.Stark",
  "Glenda.Gleichner",
  "Georgianna9",
  "Felipe35",
  "Euna86",
  "Ernesto96",
  "Ernestina0",
  "Erik.Adams93",
  "Emmitt4",
  "Emmet.Rogahn58",
  "Emerson.Nikolaus72",
  "Elroy67",
  "Elisha.Erdman",
  "Elian33",
  "Dwight90",
  "Dorthy34",
  "Dorian60",
  "Dejah.Bechtelar33",
  "Davion90",
  "Cora91",
  "Clifton.Powlowski",
  "Chad.Turcotte17",
  "Casper.Hagenes5",
  "Carson.Auer",
  "Carissa.Dooley",
  "Brittany3",
  "Bernie67",
  "Bernardo16",
  "Bartholome.Beer37",
  "Aurore68",
  "Aurore.Tillman",
  "Aurelia.Walsh",
  "Assunta61",
  "Assunta.Kautzer",
  "Antonette.Dickens0",
  "Anna99",
  "Amina2",
  "Amara.Ernser14",
  "Alicia.Lebsack68",
]
`;

exports[`Searching Profiles E2E Tests will sort results by username in desc order 2`] = `
[
  "Zora.Stroman31",
  "Yesenia.Wisozk",
  "Winifred.Feil",
  "Winfield.Ebert",
  "Velda42",
  "Van89",
  "Tyra.Rutherford",
  "Talia.Parisian94",
  "Sven.Schulist13",
  "Stephen.Goldner60",
  "Skye.Walter59",
  "Simone.Howe10",
  "Sheridan.Kulas13",
  "Shad.Daniel29",
  "Selmer68",
  "Samanta.Cummings",
  "Sallie.Schultz",
  "Roxanne58",
  "Rosamond56",
  "Reuben.Langosh45",
  "Rahul.Feil",
  "Novella.Reichel",
  "Nikki.Murphy99",
  "Nicolas.Keeling",
  "Natalie.Corwin",
  "Mellie.Berge79",
  "Maverick.Bauch14",
  "Maiya37",
  "Magali.Friesen",
  "Lolita56",
  "Lina.Hayes98",
  "Lexie76",
  "Leopoldo39",
  "Leopold59",
  "Kylie64",
  "Kristy54",
  "Kristoffer.Feest",
  "Kenna.Dare",
  "Kendrick.Lakin",
  "Kendra.Gusikowski39",
  "Kelvin5",
  "Keanu.Klocko",
  "Katrine.Kiehn69",
  "Kaitlin64",
  "Julia.Little89",
  "Judson.Halvorson41",
  "Josie.Sawayn44",
  "Josianne54",
  "Jerald.Klein37",
  "Jenifer.Mitchell92",
  "Jaunita71",
  "Jake.Hermann65",
  "Jace88",
  "Immanuel.Davis12",
  "Imelda1",
  "Houston.Mante",
  "Hortense20",
  "Hiram.Renner1",
  "Hilda.Hodkiewicz",
  "Herminio44",
  "Henriette14",
  "Golda.Stark",
  "Glenda.Gleichner",
  "Georgianna9",
  "Felipe35",
  "Euna86",
  "Ernesto96",
  "Ernestina0",
  "Erik.Adams93",
  "Emmitt4",
  "Emmet.Rogahn58",
  "Emerson.Nikolaus72",
  "Elroy67",
  "Elisha.Erdman",
  "Elian33",
  "Dwight90",
  "Dorthy34",
  "Dorian60",
  "Dejah.Bechtelar33",
  "Davion90",
  "Cora91",
  "Clifton.Powlowski",
  "Chad.Turcotte17",
  "Casper.Hagenes5",
  "Carson.Auer",
  "Carissa.Dooley",
  "Brittany3",
  "Bernie67",
  "Bernardo16",
  "Bartholome.Beer37",
  "Aurore68",
  "Aurore.Tillman",
  "Aurelia.Walsh",
  "Assunta61",
  "Assunta.Kautzer",
  "Antonette.Dickens0",
  "Anna99",
  "Amina2",
  "Amara.Ernser14",
  "Alicia.Lebsack68",
]
`;
